### PR TITLE
Initiate config management and add get config

### PR DESCRIPTION
**Pull request by StephaneTriomphe**

### CODE REVIEW

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 1]
```diff
@@ -15,8 +15,194 @@
...
+    uint16_t version[4];
+    uint8_t max_mtu;
+    uint8_t ch_range_min;
+    uint8_t ch_range_max;
+    uint8_t pdu_buffer_size;
+
+    bool CipherKeySet;
+    bool AuthenticationKeySet;
+    uint8_t StackStatus;
+    uint16_t ac_range_min_cur;  // 0 means unset ?
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:35]: Let's order them by size. Order in the list doesn't really matter.

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 2]
```diff
@@ -15,8 +15,194 @@
...
+    uint16_t ac_range_max_cur;
+
+    app_role_t node_address;
+    app_role_t app_node_role;
+    wp_NodeRole wp_node_role;  // same as app_node_role, different format
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:40]: Why is it needed?

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 3]
```diff
@@ -15,8 +15,194 @@
...
+    uint8_t ch_range_max;
+    uint8_t pdu_buffer_size;
+
+    bool CipherKeySet;
+    bool AuthenticationKeySet;
+    uint8_t StackStatus;
+    uint16_t ac_range_min_cur;  // 0 means unset ?
+    uint16_t ac_range_max_cur;
+
+    app_role_t node_address;
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:38]: Wrong type

**Review by GwendalRaoul — CHANGES_REQUESTED** [review]

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 4]
```diff
@@ -15,8 +15,194 @@
...
+    app_role_t app_node_role;
+    wp_NodeRole wp_node_role;  // same as app_node_role, different format
+    net_addr_t network_address;
+    net_channel_t network_channel;
+
+    msap_app_config_data_write_req_pl_t app_config;
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:44]: Type is a bit strange.
I would have add app_config_data that is 80 bytes, sequence that is uint8_t (even if not used anymore nowadays) and diag interval, as a dedicated field

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 5]
```diff
@@ -15,8 +15,194 @@
...
+                               "Cipher key set", false);
+    res &= get_value_from_node(WPC_is_authentication_key_set, &m_sink_config.AuthenticationKeySet, NULL,
+                               "Authentication key set", false);
+    res &= get_value_from_node(WPC_get_stack_status, &m_sink_config.StackStatus, NULL, "Stack Status", false);
+
+    res &= get_value_from_node(WPC_get_access_cycle_range, &m_sink_config.ac_range_min_cur, &m_sink_config.ac_range_max_cur,
+                               "Current access cycle range", false);
+
+    res &= get_value_from_node(WPC_get_node_address, &m_sink_config.node_address, NULL,
+                               "Network address", m_sink_config.StackStatus & APP_STACK_NODE_ADDRESS_NOT_SET);
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:154]: Why using m_sink_config.StackSatus here?

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 6]
```diff
@@ -15,8 +15,194 @@
...
+
+    res &= get_value_from_node(WPC_get_role, &role, NULL,
+                               "Node role", m_sink_config.StackStatus & APP_STACK_ROLE_NOT_SET);
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:157]: Same here

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 7]
```diff
@@ -198,6 +198,73 @@ static bool initialize_variables()
...
+void Proto_config_fill_config(wp_SinkReadConfig * config_p)
+{
+    _Static_assert( member_size(wp_AppConfigData, app_config_data)
+                    == member_size(msap_app_config_data_write_req_pl_t, app_config_data));
+
+    uint8_t status = m_sink_config.StackStatus;
+
+    *config_p = (wp_SinkReadConfig){
+        /* Sink minimal config */
+        .has_node_role = ((status & APP_STACK_ROLE_NOT_SET) ? false : true),
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:211]: Why is Node role and m_sink_config.StackStatus mixed here?

File: lib/wpc_proto/internal_modules/proto_config.c (hunk 1 above)
Comment (StephaneTriomphe) [reply lib/wpc_proto/internal_modules/proto_config.c:35]: As we don't need to pack it, I'd prefer to leave it in a defined order, but ok, it's done.

File: lib/wpc_proto/internal_modules/proto_config.c (hunk 3 above)
Comment (StephaneTriomphe) [reply lib/wpc_proto/internal_modules/proto_config.c:38]: corrected

File: lib/wpc_proto/internal_modules/proto_config.c (hunk 4 above)
Comment (StephaneTriomphe) [reply lib/wpc_proto/internal_modules/proto_config.c:44]: as discussed, removed.

File: lib/wpc_proto/internal_modules/proto_config.c (hunk 5 above)
Comment (StephaneTriomphe) [reply lib/wpc_proto/internal_modules/proto_config.c:154]: I remove all the usage of stackStatus bits (except stack on/off).

File: lib/wpc_proto/internal_modules/proto_config.c (hunk 2 above)
Comment (StephaneTriomphe) [reply lib/wpc_proto/internal_modules/proto_config.c:40]: the two version are so different that I prefered to keep the two versions. But ok, I'll change, store only one, and create conversion methods.

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 8]
```diff
@@ -15,11 +15,238 @@
...
+    {
+        proto_role->flags[flags_count] = wp_NodeRole_RoleFlags_AUTOROLE;
+        flags_count++;
+    }
+    proto_role->flags_count = flags_count;
+}
+
+static bool initialize_config_variables()
+{
+    _Static_assert(member_size(wp_NodeRole, flags) == (2 * sizeof(wp_NodeRole_RoleFlags)) );
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:107]: Not sure to understand the purpose of this test.

From proto generated file we have:
```
typedef enum _wp_NodeRole_RoleFlags {
    wp_NodeRole_RoleFlags_LOW_LATENCY = 1,
    wp_NodeRole_RoleFlags_AUTOROLE = 2
} wp_NodeRole_RoleFlags;

/* Struct definitions */
typedef struct _wp_NodeRole {
    wp_NodeRole_BaseRole role;
    pb_size_t flags_count;
    wp_NodeRole_RoleFlags flags[2];
} wp_NodeRole;
```

So you are doing a check that protoc has respected our following option hint ;-) :

`wp.NodeRole.flags max_count:2`

**Review by GwendalRaoul — APPROVED** [review]
If you could just check comment and remove check if useless, that would be nice.

Otherwise ok for me

File: lib/wpc_proto/internal_modules/proto_config.c (hunk 8 above)
Comment (GwendalRaoul) [reply lib/wpc_proto/internal_modules/proto_config.c:107]: Ok, but in this case, we have instructed the code to do it with "options". So I feel that the check is a bit overkill.
If we don't trust protoc code generator, then we have a big problem :-)

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 9]
```diff
@@ -15,11 +15,259 @@
...
+    uint8_t pdu_buffer_size;
+    uint8_t StackStatus;
+    net_channel_t network_channel;
+    app_role_t app_node_role;
+} sink_config_t;
+
+/* TODO : Protect config access */
+
+/** Sink config storage */
+sink_config_t m_sink_config;
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:45]: static

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 10]
```diff
@@ -15,11 +15,259 @@
...
+    if (res != APP_RES_OK)
+    {
+        LOGE("Cannot get %s from node\n", var_name);
+        return false;
+    }
+
     return true;
 }
 
+void convert_role_to_proto_format(app_role_t role, wp_NodeRole * proto_role)
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:88]: static

File: lib/wpc_proto/internal_modules/proto_config.c [hunk 11]
```diff
@@ -15,11 +15,259 @@
...
+    }
+}
+
+
+bool Proto_config_init(void)
+{
+    /* Read initial config from sink */
+    initialize_config_variables();
+
+    return APP_RES_PROTO_OK;
```
Comment (GwendalRaoul) [inline lib/wpc_proto/internal_modules/proto_config.c:250]: Why not propagating the result from  initialize_config_variables();

File: lib/wpc_proto/wpc_proto.c [hunk 12]
```diff
@@ -33,6 +33,80 @@
 // Max possible size of encoded message
 #define MAX_PROTOBUF_SIZE WP_CONFIG_MESSAGE_PB_H_MAX_SIZE
 
+static onEventStatus_cb_f m_onProtoEventStatus_cb = NULL;
+
+
+static void onStackStatusReceived(uint8_t status)
```
Comment (GwendalRaoul) [inline lib/wpc_proto/wpc_proto.c:39]: Why not moving directly this code to proto_config module directly?
This is where it belongs, no?

But we can do it later on

**Pull request merged**